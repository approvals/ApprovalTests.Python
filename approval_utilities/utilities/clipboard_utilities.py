def copy_to_clipboard(text: str) -> None:
    """
    This acts as a wrapper to defer the importing of pyperclip util actual usage
    """
    import pyperclip

    pyperclip.copy(text)
