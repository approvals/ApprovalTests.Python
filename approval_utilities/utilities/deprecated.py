import functools
import warnings
from typing import Callable


def deprecated(reason: str) -> Callable:
    def decorator(func: Callable) -> Callable:
        """This is a decorator which can be used to mark functions
        as deprecated. It will result in a warning being emitted
        when the function is used."""

        @functools.wraps(func)
        def new_func(*args, **kwargs):
            warnings.simplefilter("always", DeprecationWarning)  # turn off filter
            warnings.warn(
                f"Call to deprecated function {func.__name__}. {reason}",
                category=DeprecationWarning,
                stacklevel=2,
            )
            warnings.simplefilter("default", DeprecationWarning)  # reset filter
            return func(*args, **kwargs)

        return new_func

    return decorator
