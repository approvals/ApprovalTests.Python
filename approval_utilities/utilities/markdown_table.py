from typing import TYPE_CHECKING, Any, Callable, Iterable

from typing_extensions import override

from approval_utilities.approvaltests.core.verifiable import Verifiable
from approval_utilities.approvaltests.core.verify_parameters import VerifyParameters

if TYPE_CHECKING:
    from approvaltests.core.options import Options


class MarkdownTable(Verifiable):
    def __init__(self) -> None:
        self.markdown: str = ""

    @override
    def get_verify_parameters(self, options: "Options") -> VerifyParameters:
        return VerifyParameters(options.for_file.with_extension(".md"))

    @staticmethod
    def with_headers(*column_names: str) -> "MarkdownTable":
        table = MarkdownTable()
        table.add_rows(*column_names)
        dividers = map(lambda _: "---", column_names)
        table.add_rows(*dividers)
        return table

    def add_rows(self, *column_names: str) -> "MarkdownTable":
        self.markdown += MarkdownTable.print_row(*column_names)
        return self

    @override
    def __str__(self) -> str:
        return self.markdown

    @staticmethod
    def print_row(*column_names: Any) -> str:
        row = "| " + " | ".join(map(str, column_names)) + " |"
        return row + "\n"

    def add_rows_for_inputs(
        self, inputs: Iterable[Any], *input_transformers: Callable[[Any], Any]
    ) -> "MarkdownTable":
        def transform_resolver_for_input(
            input_value: Any,
        ) -> Callable[[Callable[[Any], Any]], Any]:
            return lambda transform: transform(input_value)

        for row_input in inputs:
            row = [row_input]
            row += map(transform_resolver_for_input(row_input), input_transformers)
            self.add_rows(*row)
        return self
