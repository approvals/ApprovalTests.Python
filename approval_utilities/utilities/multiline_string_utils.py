import textwrap
from typing import Optional


def remove_indentation_from(text: Optional[str]) -> str:
    if not text:
        return ""
    cleaned = textwrap.dedent(text + "|")
    if cleaned.startswith("\n"):
        cleaned = cleaned[1:]
    return cleaned[:-1]
