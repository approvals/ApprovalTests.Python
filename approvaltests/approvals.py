from threading import local
import xml.dom.minidom

from approvaltests import to_json
from approvaltests.approval_exception import ApprovalException
from approvaltests.core.namer import Namer
from approvaltests.core.scenario_namer import ScenarioNamer
from approvaltests.core.pytest_namer import PytestNamer
from approvaltests.file_approver import FileApprover
from approvaltests.list_utils import format_list
from approvaltests.reporters.diff_reporter import DiffReporter
from approvaltests.string_writer import StringWriter

DEFAULT_REPORTER = local()


def set_default_reporter(reporter):
    global DEFAULT_REPORTER
    DEFAULT_REPORTER.v = reporter
    

def get_default_reporter():
    if not hasattr(DEFAULT_REPORTER, "v") or DEFAULT_REPORTER.v is None:
        return DiffReporter()
    return DEFAULT_REPORTER.v


def get_reporter(reporter):
    return reporter or get_default_reporter()


def get_default_namer():
    return Namer()


def verify(data, reporter=None, namer=None):
    reporter_to_use = reporter or get_default_reporter()
    namer_to_use = namer or get_default_namer()
    verify_with_namer(data, namer_to_use, reporter_to_use)


def verify_with_namer(data, namer, reporter):
    reporter = get_reporter(reporter)
    approver = FileApprover()
    writer = StringWriter(data)
    error = approver.verify(namer, writer, reporter)
    if error is not None:
        raise ApprovalException(error)


def verify_as_json(object, reporter=None):
    n_ = to_json(object) + "\n"
    verify(n_, reporter)


def verify_xml(xml_string, reporter=None):
    try:
        dom = xml.dom.minidom.parseString(xml_string)
        pretty_xml = dom.toprettyxml()
    except Exception:
        pretty_xml = xml_string
    verify_with_namer(pretty_xml, Namer(extension='.xml'), reporter)


def verify_file(file_name, reporter=None):
    with open(file_name, 'r') as f:
        file_contents = f.read()
    verify(file_contents, reporter)


def verify_all(header, alist, formatter=None, reporter=None):
    text = format_list(alist, formatter, header)
    verify(text, reporter)


def get_scenario_namer(scenario_name):
    return ScenarioNamer(get_default_namer(), scenario_name)

def get_pytest_namer(pytest_request):
    return PytestNamer(get_default_namer(), pytest_request)