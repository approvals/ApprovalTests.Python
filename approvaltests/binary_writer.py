from typing import ByteString

from typing_extensions import override

from approval_utilities.utils import create_directory_if_needed
from approvaltests.core.writer import Writer


class BinaryWriter(Writer):
    contents: ByteString

    def __init__(
        self,
        contents: ByteString,
        extension: str,
    ) -> None:
        self.contents = contents
        self.extension_with_dot = extension

    @override
    def write_received_file(self, received_file: str) -> str:
        create_directory_if_needed(received_file)
        with open(received_file, mode="wb") as file:
            file.write(self.contents)

        return received_file
