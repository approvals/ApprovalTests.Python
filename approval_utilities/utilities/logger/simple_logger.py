import threading
from typing import Any, Callable, ContextManager, Optional, Union

from approval_utilities.utilities.logger.logging_instance import LoggingInstance
from approval_utilities.utilities.string_wrapper import StringWrapper
from approval_utilities.utilities.wrapper import SingleWrapper, ThreadedWrapper, Wrapper


class SimpleLogger:
    _wrapper: Wrapper[LoggingInstance] = SingleWrapper(LoggingInstance())

    @staticmethod
    def register_logger(log_method: Callable[[str], None]) -> None:
        SimpleLogger._wrapper.get().logger = log_method

    @staticmethod
    def log_to_string(log_separate_threads: bool = True) -> StringWrapper:
        with threading.Lock():
            if log_separate_threads and not isinstance(
                SimpleLogger._wrapper, ThreadedWrapper
            ):
                SimpleLogger._wrapper = ThreadedWrapper(lambda: LoggingInstance())
        return SimpleLogger._wrapper.get().log_to_string()

    @staticmethod
    def use_markers(
        parameter_text: Optional[Union[str, Callable[[], str]]] = None,
    ) -> ContextManager[None]:
        return SimpleLogger._wrapper.get().use_markers(
            parameter_text, additional_stack=1
        )

    @staticmethod
    def variable(name: str, value: Any, show_types: bool = False) -> None:
        SimpleLogger._wrapper.get().variable(name, value, show_types=show_types)

    @staticmethod
    def hour_glass() -> None:
        SimpleLogger._wrapper.get().hour_glass()

    @staticmethod
    def event(event_name: str) -> None:
        SimpleLogger._wrapper.get().event(event_name)

    @staticmethod
    def show_timestamps(display: bool) -> None:
        SimpleLogger._wrapper.get().log_with_timestamps = display

    @staticmethod
    def query(query_text: str) -> None:
        SimpleLogger._wrapper.get().query(query_text)

    @staticmethod
    def message(message: str) -> None:
        SimpleLogger._wrapper.get().message(message)

    @staticmethod
    def warning(
        text: Union[str, BaseException] = "", exception: Optional[BaseException] = None
    ) -> None:
        SimpleLogger._wrapper.get().warning(text, exception)

    @staticmethod
    def show_queries(show: bool) -> None:
        SimpleLogger._wrapper.get().show_queries(show)

    @staticmethod
    def show_all(show: bool) -> None:
        SimpleLogger._wrapper.get().show_all(show)

    @staticmethod
    def show_messages(show: bool) -> None:
        SimpleLogger._wrapper.get().show_messages(show)

    @staticmethod
    def show_variables(show: bool) -> None:
        SimpleLogger._wrapper.get().show_variables(show)

    @staticmethod
    def show_hour_glass(show: bool) -> None:
        SimpleLogger._wrapper.get().show_hour_glass(show)

    @staticmethod
    def show_markers(show: bool) -> None:
        SimpleLogger._wrapper.get().show_markers(show)

    @staticmethod
    def show_events(show: bool) -> None:
        SimpleLogger._wrapper.get().show_events(show)
