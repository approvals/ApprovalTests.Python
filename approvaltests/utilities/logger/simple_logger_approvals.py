from types import TracebackType
from typing import ContextManager, Optional, Type

from approval_utilities.utilities.logger.simple_logger import SimpleLogger
from approvaltests.approvals import verify
from approvaltests.core.options import Options


def verify_simple_logger(
    *,  # enforce keyword arguments - https://www.python.org/dev/peps/pep-3102/
    options: Optional[Options] = None,
) -> ContextManager[None]:
    class VerifySimpleLogger:
        def __init__(self) -> None:
            self.output = SimpleLogger.log_to_string()

        def __enter__(self) -> None:
            pass

        def __exit__(
            self,
            exc_type: Optional[Type[BaseException]],
            exc_val: Optional[BaseException],
            exc_tb: Optional[TracebackType],
        ) -> bool:
            verify(self.output, options=options)

    return VerifySimpleLogger()
