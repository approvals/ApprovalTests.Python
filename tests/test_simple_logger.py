import datetime
import re
import threading
from contextlib import contextmanager
from typing import Callable, Generator

from approval_utilities.utilities.logger.simple_logger import SimpleLogger
from approval_utilities.utilities.time_utilities import use_utc_timezone
from approvaltests import (
    Options,
    approvals,
    run_all_combinations,
    verify,
    verify_logging_for_all_combinations,
)
from approvaltests.utilities.logger.simple_logger_approvals import verify_simple_logger


def test_warnings() -> None:
    def scrubber(text: str) -> str:
        return re.sub(
            re.escape(__file__), "test_simple_logger.py", text, flags=re.IGNORECASE
        )

    output = SimpleLogger.log_to_string()
    SimpleLogger._wrapper.get().log_stack_traces = True
    text = "EVERYTHING IS AWFUL!!!!!!"
    try:
        raise Exception("EVERYTHING IS exceptionally AWFUL!!!!!!")
    except Exception as e:
        exception = e
    SimpleLogger.warning(text)
    SimpleLogger.warning(exception)
    SimpleLogger.warning(text, exception)
    verify(output, options=Options().with_scrubber(scrubber))


def log_from_inner_method() -> None:
    with SimpleLogger.use_markers():
        name = "Example"
        SimpleLogger.variable("name", name)
        for _ in range(0, 142):
            SimpleLogger.hour_glass()


def test_standard_logger() -> None:
    with verify_simple_logger():
        with SimpleLogger.use_markers() as m:
            log_from_inner_method()


def test_timestamps() -> None:
    with use_utc_timezone():
        with verify_simple_logger():
            count = -1

            def create_applesauce_timer() -> datetime.datetime:
                dates = [
                    datetime.datetime.fromtimestamp(0.0),
                    datetime.datetime.fromtimestamp(0.5),
                    datetime.datetime.fromtimestamp(2.0),
                    datetime.datetime.fromtimestamp(1050),
                    datetime.datetime.fromtimestamp(1052),
                ]
                nonlocal count
                count = count + 1
                return dates[count]

            SimpleLogger._wrapper.get().timer = create_applesauce_timer
            SimpleLogger.show_timestamps(True)
            SimpleLogger.event("1")
            SimpleLogger.event("2")
            SimpleLogger.event("3")
            SimpleLogger.event("4")
            SimpleLogger.warning(exception=Exception("Oh no you didn't!"))


# begin-snippet: verify_simple_logger_example
def test_variable() -> None:
    with verify_simple_logger():
        SimpleLogger.variable("dalmatians", 101, show_types=True)
        SimpleLogger.variable("dalmatians", 101, show_types=False)


# end-snippet


# begin-snippet: verify_simple_logger_long_example
def test_variable_explict() -> None:
    output = SimpleLogger.log_to_string()
    SimpleLogger.variable("dalmatians", 101, show_types=True)
    SimpleLogger.variable("dalmatians", 101, show_types=False)
    verify(output)


# end-snippet


def test_variable_with_list() -> None:
    output = SimpleLogger.log_to_string()
    with SimpleLogger.use_markers():
        names = ["Jacqueline", "Llewellyn"]
        SimpleLogger.variable("names", names, show_types=True)
        SimpleLogger.variable("names", names, show_types=False)
    verify(output)


def verify_toggle(toggle_name: str, toggle: Callable[[bool], None]) -> None:
    SimpleLogger.show_all(True)
    SimpleLogger.event(f"Toggle Off {toggle_name}")
    toggle(False)
    log_everything()


def test_switching() -> None:
    output = SimpleLogger.log_to_string()

    verify_toggle("None", lambda a: SimpleLogger.show_all(True)),
    verify_toggle("All", SimpleLogger.show_all),
    verify_toggle("Query", SimpleLogger.show_queries),
    verify_toggle("Message", SimpleLogger.show_messages),
    verify_toggle("Variable", SimpleLogger.show_variables),
    verify_toggle("Hour Glass", SimpleLogger.show_hour_glass),
    verify_toggle("Markers", SimpleLogger.show_markers),
    verify_toggle("Events", SimpleLogger.show_events),

    verify(output)


def log_everything() -> None:
    with SimpleLogger.use_markers():
        SimpleLogger.query("Select * from people")
        SimpleLogger.variable("Nonsense", "foo")
        SimpleLogger.event("Testing")
        SimpleLogger.message("Something random")
        for a in range(1, 13):
            SimpleLogger.hour_glass()
        try:
            infinity = 1 / 0
        except Exception as e:
            SimpleLogger.warning(exception=e)


def function_to_run(color: str, number: str) -> None:
    with SimpleLogger.use_markers():
        SimpleLogger.variable("color", color)
        SimpleLogger.variable("number", number)
        if number == "brie":
            raise Exception("AHHHHHH!")


def test_use_markers_with_raised_exception() -> None:
    def throw_exception() -> None:
        with SimpleLogger.use_markers():
            raise Exception("Everything is awflu!!?")

    output = SimpleLogger.log_to_string()
    try:
        throw_exception()
    except BaseException as e:
        SimpleLogger.warning(e)
    verify(output)


def test_run_combinations() -> None:
    with verify_simple_logger():
        run_all_combinations(function_to_run, [["red", "blue"], ["one", "two", "brie"]])


def test_run_combinations_with_exception_handler() -> None:
    approvals.settings().allow_multiple_verify_calls_for_this_method()

    with verify_simple_logger():
        run_all_combinations(
            function_to_run,
            [["red"], ["one", "brie"]],
            exception_handler=lambda e: SimpleLogger.warning(exception=e),
        )
    with verify_simple_logger():
        run_all_combinations(
            function_to_run,
            [["red"], ["one", "brie"]],
            exception_handler=lambda e: SimpleLogger.warning(e),
        )
    with verify_simple_logger():
        run_all_combinations(
            function_to_run,
            [["red"], ["one", "brie"]],
            exception_handler=SimpleLogger.warning,
        )


def test_verify_logging_for_all_combinations() -> None:
    verify_logging_for_all_combinations(
        function_to_run, [["red", "blue"], ["one", "two", "brie"]]
    )


# begin-snippet: method_with_inputs
def method_with_inputs(number: int, name: str) -> None:
    with SimpleLogger.use_markers(f"number = {number}, name = {name}"):
        # end-snippet
        print(f"{number}) {name}")


def test_markers_with_signature() -> None:
    output = SimpleLogger.log_to_string()
    method_with_inputs(1, "Susan")
    verify(output)


# begin-snippet: method_with_inputs_and_outputs
def method_with_inputs_and_outputs(number: int, announcement: str) -> None:
    with SimpleLogger.use_markers(
        lambda: f"number = {number}, announcement = {announcement}"
    ):
        # end-snippet
        for number in range(number, 0, -1):
            print(number)
        print(announcement)


def test_markers_with_signature_in_and_out() -> None:
    output = SimpleLogger.log_to_string()
    method_with_inputs_and_outputs(10, "Blast off")
    verify(output)


def test_race_condition() -> None:
    whose_turn = 0

    @contextmanager
    def wait_for(number: int) -> Generator[None, None, None]:
        nonlocal whose_turn
        while whose_turn < number:
            pass
        yield
        whose_turn += 1

    log1 = "Log1"

    def thread_1() -> None:
        nonlocal log1
        with wait_for(0):
            log1 = SimpleLogger.log_to_string(True)
        with wait_for(2):
            SimpleLogger.event("event_a")
        with wait_for(4):
            SimpleLogger.event("event_b")

    log2 = "Log2"

    def thread_2() -> None:
        nonlocal log2
        with wait_for(1):
            log2 = SimpleLogger.log_to_string(True)
        with wait_for(3):
            SimpleLogger.event("event_a")
        with wait_for(5):
            SimpleLogger.event("event_b")

    threading.Thread(target=thread_1).start()
    threading.Thread(target=thread_2).start()

    with wait_for(6):
        assert f"'{log1}'" == f"'{log2}'"
