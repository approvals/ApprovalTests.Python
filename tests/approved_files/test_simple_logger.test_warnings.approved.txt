*******************************************************************************************
Message:EVERYTHING IS AWFUL!!!!!!
*******************************************************************************************
*******************************************************************************************
Traceback (most recent call last):
  File "test_simple_logger.py", line 27, in test_warnings
    raise Exception("EVERYTHING IS exceptionally AWFUL!!!!!!")
Exception: EVERYTHING IS exceptionally AWFUL!!!!!!

*******************************************************************************************
*******************************************************************************************
Message:EVERYTHING IS AWFUL!!!!!!
Traceback (most recent call last):
  File "test_simple_logger.py", line 27, in test_warnings
    raise Exception("EVERYTHING IS exceptionally AWFUL!!!!!!")
Exception: EVERYTHING IS exceptionally AWFUL!!!!!!

*******************************************************************************************
